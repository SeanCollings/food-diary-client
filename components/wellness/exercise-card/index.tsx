import TextArea from '@components/ui/text-area';
import { MEDIA_DESKTOP, MEDIA_MOBILE } from '@utils/constants';
import Image from 'next/image';
import { ChangeEvent, FC, useEffect, useState } from 'react';
import { useDateSelectedContext } from '@store/date-selected-context';
import styled from 'styled-components';
import {
  TExcercise,
  useWellnessEntriesContext,
} from '@store/wellness-entries-context';
import TimeInputCustom from '@components/ui/input/time-input-custom';
import { runValidations } from '@utils/validation';
import { excerciseValidators } from '@utils/validation/validators/collections';
import { trim } from '@utils/string-utils';
import { TExcerciseEntry } from '@lib/interfaces/wellness.interface';

const SContainer = styled.div`
  padding: 20px;
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: center;
  flex-wrap: wrap;

  ${MEDIA_DESKTOP} {
    gap: 20px;
    flex-wrap: nowrap;
  }
`;
const STitleContainer = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
`;
const SAllDetailsContainer = styled.div`
  width: 100%;
  display: flex;
  flex-direction: column;
  align-items: baseline;

  ${MEDIA_MOBILE} {
    align-items: center;
  }
`;
const STimeInputContainer = styled.div``;
const STextAreaContainer = styled.div`
  width: 100%;
  justify-content: center;
  display: flex;
`;
const SHeader = styled.div`
  padding-bottom: 12px;
  user-select: none;
  text-align: center;
`;
const SImageContainer = styled.div`
  height: 80px;
  width: 80px;
  user-select: none;
  margin-bottom: 8px;
  position: relative;
  text-align: center;
`;
const SImage = styled(Image)`
  width: auto;
  height: 100%;
  max-width: 100%;
`;

interface IRunFormValidations {
  details: string;
  time: string;
}

const runFormValidations = (values: IRunFormValidations) =>
  runValidations([
    excerciseValidators['details']({
      value: values.details,
    }),
    excerciseValidators['time']({
      value: values.time,
    }),
  ]);

interface IProps {
  id: string;
  title: string;
  imageSrc: string;
}

const ExcerciseCard: FC<IProps> = ({ title, imageSrc }) => {
  const { dateSelectedISO } = useDateSelectedContext();
  const { wellnessEntries, updateEntryByKey } = useWellnessEntriesContext();

  const [hasUpdated, setHasUpdated] = useState(false);
  const [hasBlurred, setHasBlurred] = useState(false);
  const [time, setTime] = useState('00:00');
  const [excerciseDetails, setExcerciseDetails] = useState('');
  const [errors, setErrors] = useState<Partial<IRunFormValidations>>({});

  useEffect(() => {
    const currentDayExcerciseEntry = (
      wellnessEntries[dateSelectedISO] as TExcerciseEntry
    )?.excercise;

    setTime(currentDayExcerciseEntry?.time || '00:00');
    setExcerciseDetails(currentDayExcerciseEntry?.details || '');
  }, [dateSelectedISO, wellnessEntries]);

  useEffect(() => {
    const listener = () => {
      updateEntryByKey<TExcercise>({
        date: dateSelectedISO,
        type: 'excercise',
        content: { time, details: trim(excerciseDetails) },
      });
      setHasUpdated(false);
      setHasBlurred(false);
    };

    if (hasUpdated && hasBlurred) {
      const errors = runFormValidations({ details: excerciseDetails, time });

      if (!!Object.keys(errors).length) {
        return setErrors(errors);
      }

      listener();
    }
  }, [
    time,
    excerciseDetails,
    hasBlurred,
    hasUpdated,
    dateSelectedISO,
    updateEntryByKey,
  ]);

  const handleDetailsChange = (event: ChangeEvent<HTMLTextAreaElement>) => {
    setExcerciseDetails(event.target.value);
    if (!hasUpdated) setHasUpdated(true);
  };
  const handleTimeChange = (value: string) => {
    if (!value) {
      return;
    }

    if (time !== value) {
      setTime(value);
      setHasUpdated(true);
    }
  };
  const handleOnBlur = () => {
    if (!hasBlurred) setHasBlurred(true);

    const errors = runFormValidations({ details: excerciseDetails, time });
    setErrors(errors);
  };

  return (
    <SContainer>
      <STitleContainer>
        <SHeader>{title}</SHeader>
        <SImageContainer>
          <SImage
            src={imageSrc}
            alt={title}
            width={128}
            height={180}
            priority={false}
          />
        </SImageContainer>
      </STitleContainer>
      <SAllDetailsContainer>
        <STextAreaContainer>
          <TextArea
            id="exercise"
            minWidth={250}
            value={excerciseDetails}
            isError={errors.details}
            borderStyle={'dashed'}
            borderWidth={1}
            totalRows={3}
            placeholder="..."
            title="add excercise notes"
            onChange={handleDetailsChange}
            onBlur={handleOnBlur}
          />
        </STextAreaContainer>
        <STimeInputContainer>
          <TimeInputCustom
            id="exercise_time"
            value={time}
            isError={errors.time}
            onChange={handleTimeChange}
            onBlur={handleOnBlur}
          />
        </STimeInputContainer>
      </SAllDetailsContainer>
    </SContainer>
  );
};

export default ExcerciseCard;
