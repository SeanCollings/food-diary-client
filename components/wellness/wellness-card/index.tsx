import { FC, useEffect, useRef, useState } from 'react';
import styled from 'styled-components';
import Image from 'next/image';
import { MdAddCircle, MdRemoveCircle } from 'react-icons/md';
import { TWellnessTypes } from '@utils/interfaces';
import { useDateSelectedContext } from '@store/date-selected-context';
import {
  TDrink,
  useWellnessEntriesContext,
} from '@store/wellness-entries-context';
import { COLOURS } from '@utils/constants';

interface SCount {
  hasNoValue: boolean;
}

const SContainer = styled.div`
  padding: 20px;
  display: flex;
  flex-direction: column;
  align-items: center;
`;
const SHeader = styled.div`
  padding-bottom: 12px;
  text-align: center;
`;
const SImageContainer = styled.div`
  height: 50px;
  width: 50px;
  user-select: none;
  cursor: pointer;

  :hover {
    filter: drop-shadow(0px 0px 1px ${COLOURS.gray_dark});
  }
`;
const SCounterContainer = styled.div`
  padding-top: 12px;
  font-size: 20px;
  display: flex;
  align-items: center;
  user-select: none;
`;
const SCount = styled.div<SCount>`
  transition: 0.1s;
  text-align: center;
  min-width: 54px;

  ${({ hasNoValue }) => hasNoValue && 'opacity: 0.4'};
`;
const SIconRemove = styled(MdRemoveCircle)`
  cursor: pointer;
  :hover {
    scale: 1.1;
    opacity: 0.7;
  }
`;
const SIconAdd = styled(MdAddCircle)`
  cursor: pointer;
  :hover {
    scale: 1.1;
    opacity: 0.7;
  }
`;

interface IProps {
  id: TWellnessTypes;
  title: string;
  imageSrc: string;
  color: string;
}

const WellnessCard: FC<IProps> = ({ id, title, imageSrc, color }) => {
  const { dateSelectedISO } = useDateSelectedContext();
  const { wellnessEntries, updateEntryByKey } = useWellnessEntriesContext();

  const [counter, setCounter] = useState(0);
  const [hasUpdated, setHasUpdated] = useState(false);

  useEffect(() => {
    const currentEntry = wellnessEntries[dateSelectedISO];
    setCounter((currentEntry?.[id] as TDrink)?.value || 0);
  }, [dateSelectedISO, wellnessEntries, id]);

  useEffect(() => {
    if (hasUpdated) {
      updateEntryByKey<TDrink>({
        date: dateSelectedISO,
        type: id,
        content: { value: counter },
      });
      setHasUpdated(false);
    }
  }, [id, counter, hasUpdated, dateSelectedISO, updateEntryByKey]);

  const onClickRemove = () => {
    if (counter > 0) {
      setCounter((curr) => curr - 1);
      setHasUpdated(true);
    }
  };
  const onClickAdd = () => {
    if (counter < 99) {
      setCounter((curr) => curr + 1);
      setHasUpdated(true);
    }
  };

  return (
    <SContainer>
      <SHeader>{title}</SHeader>
      <SImageContainer onClick={onClickAdd}>
        <Image
          src={imageSrc}
          alt={title}
          width="100%"
          height="100%"
          objectFit="contain"
          draggable={false}
        />
      </SImageContainer>
      <SCounterContainer>
        <SIconRemove size={24} color={color} onClick={onClickRemove} />
        <SCount hasNoValue={counter === 0}>{counter}</SCount>
        <SIconAdd size={24} color={color} onClick={onClickAdd} />
      </SCounterContainer>
    </SContainer>
  );
};

export default WellnessCard;
