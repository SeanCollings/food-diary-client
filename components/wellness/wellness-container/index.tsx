import { ThemeBorderBottom } from '@components/ui/style-themed';
import ExcerciseCard from '@components/wellness/exercise-card';
import WellnessCard from '@components/wellness/wellness-card';
import {
  APP_THEME_DEFAULT,
  MEDIA_DESKTOP,
  MEDIA_MOBILE,
} from '@utils/app.constants';
import { FC } from 'react';
import styled from 'styled-components';

const SContainer = styled.section`
  margin-top: 20px;
  display: flex;
  gap: 20px;
  flex-direction: column;

  ${MEDIA_DESKTOP} {
    flex-direction: row;
  }
  ${MEDIA_MOBILE} {
    margin-top: 0px;
  }
`;
const SDrinkContainer = styled.div`
  border-radius: 6px;
  background: white;
  display: flex;
  flex-basis: 60%;
  justify-content: space-evenly;
  flex-wrap: wrap;
  background-color: var(--bg-secondary);
  ${ThemeBorderBottom}
`;
const SExcerciseContainer = styled.div`
  border-radius: 6px;
  background: white;
  flex-basis: 40%;
  background-color: var(--bg-secondary);
  ${ThemeBorderBottom}
`;

const WellnessContainer: FC = () => {
  return (
    <SContainer>
      <SDrinkContainer>
        <WellnessCard
          id="water"
          title="Water"
          imageSrc={'/static/images/water_bottle.webp'}
          color={APP_THEME_DEFAULT.quaternary}
          width={69}
          height={193}
        />
        <WellnessCard
          id="tea_coffee"
          title="Tea / coffee"
          imageSrc={'/static/images/coffee.webp'}
          color={APP_THEME_DEFAULT.quaternary}
          width={384}
          height={348}
        />
        <WellnessCard
          id="alcohol"
          title="Alcohol"
          imageSrc={'/static/images/wine.webp'}
          color={APP_THEME_DEFAULT.quaternary}
          width={115}
          height={298}
        />
      </SDrinkContainer>
      <SExcerciseContainer>
        <ExcerciseCard
          id="alcohol_counter"
          title="Excercise"
          imageSrc={'/static/images/exercise.webp'}
        />
      </SExcerciseContainer>
    </SContainer>
  );
};

export default WellnessContainer;
