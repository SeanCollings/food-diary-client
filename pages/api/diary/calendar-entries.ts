import {
  formatMonthNumberYear,
  getDayFromDate,
  getMidnightISODaysInMonth,
  getMonthAndYearFromDate,
  getNewMonth,
  setDateMidnightISOString,
} from '@utils/date-utils';
import { NextApiRequest, NextApiResponse } from 'next';
import { diaryMealsMockData, diaryWellnessMockData } from '@client/mock';
import { URI_DIARY_CALENDAR_ENTRIES } from '@client/constants';
import { ICalendarEntriesResponseBody } from '@client/interfaces/diary-data';

interface IRequest extends NextApiRequest {
  query: {
    date: string;
    months: string;
  };
}

interface IEntriesPerMonth {
  [date: string]: number[];
}

const getDatesOfEntriesInDateRange = <T extends unknown>(
  dateRange: string[],
  entries: { [date: string]: T }
) => {
  const entryDates = dateRange.reduce((acc, date) => {
    if (entries[date]) {
      acc.push(getDayFromDate(date));
    }

    return acc;
  }, [] as number[]);

  return entryDates;
};

const getEntriesPerMonth = (monthRange: {
  [date: string]: boolean;
}): IEntriesPerMonth => {
  const entries = Object.keys(monthRange).reduce((acc, date) => {
    const [month, year] = getMonthAndYearFromDate(date);
    const datesInMonth = getMidnightISODaysInMonth(month, year);

    const mealEntryDays = getDatesOfEntriesInDateRange(
      datesInMonth,
      diaryMealsMockData
    );
    const wellnessEntryDays = getDatesOfEntriesInDateRange(
      datesInMonth,
      diaryWellnessMockData
    );

    const combinedEntryDays = mealEntryDays.concat(wellnessEntryDays);
    const uniqueDays = [...new Set(combinedEntryDays)];

    if (uniqueDays.length) {
      acc[formatMonthNumberYear(date)] = uniqueDays;
    }

    return acc;
  }, {} as IEntriesPerMonth);

  return entries;
};

const getPreviousMonthsRange = (date: string, months: number) => {
  let previousMonth = new Date(date);
  let monthRange = months;
  const formattedMonthRange: { [date: string]: boolean } = {
    [formatMonthNumberYear(date)]: true,
  };
  const monthRangeISO: { [date: string]: boolean } = {
    [setDateMidnightISOString(date)]: true,
  };

  while (monthRange > 1) {
    previousMonth = getNewMonth(previousMonth, 'previous');
    formattedMonthRange[formatMonthNumberYear(previousMonth)] = true;
    monthRangeISO[setDateMidnightISOString(previousMonth)] = true;
    monthRange -= 1;
  }

  return [formattedMonthRange, monthRangeISO];
};

const getDataFromAPI = async (date: string, months: string) => {
  const [monthRange, monthRangeISO] = getPreviousMonthsRange(
    date,
    parseInt(months)
  );
  const entries = getEntriesPerMonth(monthRangeISO);

  return Promise.resolve({ monthRange, entries });
};

const handler = async (
  req: IRequest,
  res: NextApiResponse<ICalendarEntriesResponseBody>
) => {
  const { date, months } = req.query;
  console.log(`-------- ${URI_DIARY_CALENDAR_ENTRIES} :`, date, months);

  if (req.method !== 'GET') {
    return;
  }

  const { monthRange, entries } = await getDataFromAPI(date, months);

  return res.status(200).json({ months: monthRange, entries });
};

export default handler;
