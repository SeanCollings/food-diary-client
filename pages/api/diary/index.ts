import { NextApiRequest, NextApiResponse } from 'next';
import { URI_DIARY } from '@client/constants';
import { IDiaryReponseData } from '@client/interfaces/diary-data';
import axios, { AxiosError } from 'axios';
import { getSession } from 'next-auth/react';

interface IRequest extends NextApiRequest {
  query: {
    date: string;
  };
}

const handler = async (
  req: IRequest,
  res: NextApiResponse<IDiaryReponseData>
) => {
  const session = await getSession({ req });
  const { date } = req.query;
  console.log(`-------- ${URI_DIARY} :`, date);

  if (req.method !== 'GET') {
    return res.status(500).json({ ok: false });
  }

  try {
    const { data } = await axios.get(
      `${process.env.SERVER_HOST}/diary?date=${date}`,
      {
        headers: {
          Authorization: `Bearer ${session?.accessToken}`,
        },
      }
    );

    return res.status(200).json({
      ok: true,
      entry: {
        meals: data?.meals || {},
        wellness: data?.wellness || {},
      },
    });
  } catch (err) {
    return res.status(500).json({
      ok: false,
      message: (
        err as AxiosError<{
          statusCode: number;
          message: string;
          error: string;
        }>
      ).response?.data.message,
    });
  }
};

export default handler;
