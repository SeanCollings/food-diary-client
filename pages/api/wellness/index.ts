import { CustomAxiosError } from '@client/interfaces/axios.types';
import {
  IWellnessEntriesDto,
  TWellnessEntryAndDate,
} from '@lib/interfaces/wellness.interface';
import { createApiClientSecure } from '@server/api-client';
import { API_WELLNESS } from '@server/server.constants';
import { NextApiRequest, NextApiResponse } from 'next';

interface IRequest<T extends unknown> extends NextApiRequest {
  body: T;
}
interface IResponse {
  ok: boolean;
  message?: string;
}

const handler = async (
  req: IRequest<IWellnessEntriesDto>,
  res: NextApiResponse<IResponse>,
) => {
  try {
    const apiClientSecure = await createApiClientSecure(req);

    if (req.method === 'PUT') {
      const transformedBody = Object.values(req.body).reduce(
        (acc, entry) => [...acc, entry],
        [] as TWellnessEntryAndDate[],
      );

      const { data } = await apiClientSecure.put(API_WELLNESS, {
        data: transformedBody,
      });

      return res.status(201).json({ ok: true });
    }
  } catch (err) {
    const typedError = err as CustomAxiosError;

    return res.status(typedError.status || 500).json({
      ok: false,
      message: typedError.response?.data.message,
    });
  }

  return;
};

export default handler;
