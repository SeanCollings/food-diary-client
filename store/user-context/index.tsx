import { getStructuredClone } from '@utils/get-structured-clone';
import {
  createContext,
  FC,
  ReactNode,
  useCallback,
  useContext,
  useMemo,
  useState,
  useEffect,
} from 'react';

export enum EUser {
  ID = 'id',
  NAME = 'name',
  EMAIL = 'email',
  AVATAR = 'avatar',
  SHARE_LINK = 'shareLink',
  STATS = 'stats',
  PREFERENCES = 'preferences',
}
export enum EPrefences {
  SHOW_DAY_STREAK = 'showDayStreak',
  SHOW_WEEKLY_EXERCISE = 'showWeeklyExcercise',
  SHOW_WEEKLY_WATER = 'showWeeklyWater',
  IS_PROFILE_SHARED = 'isProfileShared',
}

export interface IUserModel {
  [EUser.ID]: string;
  [EUser.NAME]: string;
  [EUser.EMAIL]: string;
  [EUser.AVATAR]?: string;
  [EUser.SHARE_LINK]?: string;
  [EUser.PREFERENCES]?: {
    [EPrefences.SHOW_DAY_STREAK]?: boolean;
    [EPrefences.SHOW_WEEKLY_EXERCISE]?: boolean;
    [EPrefences.SHOW_WEEKLY_WATER]?: boolean;
    [EPrefences.IS_PROFILE_SHARED]?: boolean;
  };
  [EUser.STATS]?: {
    dayStreak?: number;
    weeklyExercise?: string;
    weeklyWater?: number;
  };
}
export interface IPreferenceModel {}

type IPartialUserUpdate =
  | { [EUser.NAME]: string }
  | { [EUser.EMAIL]: string }
  | { [EUser.AVATAR]: string }
  | { [EUser.SHARE_LINK]: string };

type IPartialPreference =
  | { [EPrefences.SHOW_DAY_STREAK]?: boolean }
  | { [EPrefences.SHOW_WEEKLY_EXERCISE]: boolean }
  | { [EPrefences.SHOW_WEEKLY_WATER]: boolean }
  | { [EPrefences.IS_PROFILE_SHARED]: boolean };

export interface IUserContext {
  user: IUserModel | null;
  setInitialUser: (user: IUserModel) => void;
  updateUser: (updated: IPartialUserUpdate) => void;
  updatePreferences: (preferences: IPartialPreference) => void;
}

const initialState: IUserContext = {
  user: null,
  // user: {
  //   id: '',
  //   name: '',
  //   email: '',
  //   avatar: '',
  //   shareLink: '',
  //   darkMode: false,
  //   theme: {},
  //   preferences: {
  //     showDayStreak: true,
  //   },
  //   stats: {
  //     dayStreak: 5,
  //     weeklyExercise: '03:25',
  //     weeklyWater: 32,
  //   },
  // },
  setInitialUser: () => null,
  updateUser: () => null,
  updatePreferences: () => null,
};

const UserContext = createContext(initialState);

export const UserContextProvider: FC<{
  children: ReactNode;
  initialState?: Partial<IUserContext>;
}> = ({ children, initialState }) => {
  // console.log('initialState ::', initialState?.user);
  const [user, setCurrentUser] = useState<IUserModel | null>(
    initialState?.user || null
  );
  const [updatedUserFields, setUpdatedUserFields] = useState<
    IPartialUserUpdate[]
  >([]);
  const [updatedPreferences, setUpdatedPreferences] = useState<
    IPartialPreference[]
  >([]);

  useEffect(() => {
    if (!updatedUserFields.length) {
      return;
    }

    const timer = setTimeout(async () => {
      const payload = updatedUserFields.reduce((curr, update) => {
        return { ...curr, ...update };
      }, {} as IPartialUserUpdate);

      console.log('USER PATCH:', payload);

      setUpdatedUserFields([]);
    }, 500);

    return () => {
      clearTimeout(timer);
    };
  }, [updatedUserFields]);

  useEffect(() => {
    if (!updatedPreferences.length) {
      return;
    }

    const timer = setTimeout(async () => {
      const payload = updatedPreferences.reduce((curr, preference) => {
        return { ...curr, ...preference };
      }, {} as IPartialPreference);

      console.log('PREFERENCES PATCH:', payload);

      setUpdatedPreferences([]);
    }, 500);

    return () => {
      clearTimeout(timer);
    };
  }, [updatedPreferences]);

  const setInitialUser = useCallback((unitialUser: IUserModel) => {
    setCurrentUser(unitialUser);
  }, []);

  const updateUser = useCallback((update: IPartialUserUpdate) => {
    setCurrentUser((curr) => {
      if (!curr) {
        return null;
      }

      return {
        ...(getStructuredClone(curr || {}) || {}),
        ...(curr ? update : {}),
      };
    });

    setUpdatedUserFields((curr) => [...curr, update]);
  }, []);

  const updatePreferences = useCallback((preference: IPartialPreference) => {
    setCurrentUser((curr) => {
      if (!curr) {
        return null;
      }

      return {
        ...getStructuredClone(curr),
        preferences: { ...getStructuredClone(curr.preferences), ...preference },
      };
    });
    setUpdatedPreferences((curr) => [...curr, preference]);
  }, []);

  const context = useMemo(
    () => ({
      user,
      setInitialUser,
      updateUser,
      updatePreferences,
    }),
    [user, setInitialUser, updateUser, updatePreferences]
  );

  return (
    <UserContext.Provider value={context}>{children}</UserContext.Provider>
  );
};

export const useUserContext = () => useContext<IUserContext>(UserContext);
