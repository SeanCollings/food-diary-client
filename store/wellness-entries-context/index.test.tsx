import { diaryService } from '@client/services/diary.service';
import { TDrink, WellnessEntriesContextProvider, initialState } from '.';
import { act, render, waitFor } from '@testing-library/react';
import { FC, useEffect, useRef } from 'react';
import { IWellnessEntries } from '@lib/interfaces/wellness.interface';
import { useUserContext } from '@store/user-context';
import { useWellnessEntriesContext } from '.';

jest.useFakeTimers();
jest.mock('@client/services/diary.service');
jest.mock('@store/user-context');

const DisplayEntry: FC<{
  entries: IWellnessEntries;
  wellnessUpdated: boolean;
}> = ({ entries, wellnessUpdated }) => (
  <>
    {Object.keys(entries).map((key) => (
      <div key={key}>
        Key: {key}
        <div>{JSON.stringify(entries[key])}</div>
      </div>
    ))}
    <div>wellnessUpdated: {wellnessUpdated.toString()}</div>
  </>
);

const TestComponent = () => <div>Test component</div>;

const MockInitialEntries: FC<{ setEntries?: boolean }> = ({ setEntries }) => {
  const { wellnessEntries, wellnessUpdated, requestSetWellnessEntries } =
    useWellnessEntriesContext();

  useEffect(() => {
    if (setEntries) {
      requestSetWellnessEntries({
        wellness: {
          '01-02-2023': {
            water: { value: 5 },
            tea_coffee: { value: 2 },
            alcohol: { value: 1 },
            excercise: { time: '00:15', details: 'Run and walk' },
          },
        },
      });
    }
  }, [requestSetWellnessEntries, setEntries]);

  return (
    <DisplayEntry entries={wellnessEntries} wellnessUpdated={wellnessUpdated} />
  );
};

const MockUpdateEntry: FC<{
  runBothUpdates?: boolean;
  runResetWellnessUpdated?: boolean;
}> = ({ runBothUpdates = false, runResetWellnessUpdated = false }) => {
  const runRef = useRef(false);
  const {
    wellnessEntries,
    wellnessUpdated,
    updateEntryByKey,
    resetWellnessUpdated,
  } = useWellnessEntriesContext();

  useEffect(() => {
    if (runRef.current) {
      return;
    }

    runRef.current = true;

    updateEntryByKey<TDrink>({
      date: '01-02-2023',
      type: 'water',
      content: { value: 4 },
    });
    if (runBothUpdates) {
      updateEntryByKey<TDrink>({
        date: '01-02-2023',
        type: 'alcohol',
        content: { value: 7 },
      });
    }
  }, [runBothUpdates, updateEntryByKey]);

  useEffect(() => {
    if (Object.entries(wellnessEntries).length && runResetWellnessUpdated) {
      resetWellnessUpdated();
    }
  }, [wellnessEntries, runResetWellnessUpdated, resetWellnessUpdated]);

  return (
    <DisplayEntry entries={wellnessEntries} wellnessUpdated={wellnessUpdated} />
  );
};

describe('store - wellness-entries-context', () => {
  const mockDiaryService = jest.mocked(diaryService);
  const mockUseUserContext = jest.mocked(useUserContext);
  const mockError = jest.fn();

  beforeAll(() => {
    jest.spyOn(console, 'error').mockImplementation(mockError);
  });

  beforeEach(() => {
    mockDiaryService.updateWellnessEntries.mockResolvedValue({});
    mockUseUserContext.mockReturnValue({ userLoggedIn: false } as any);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should return initial state', () => {
    const {
      wellnessEntries,
      wellnessUpdated,
      requestSetWellnessEntries,
      updateEntryByKey,
      resetWellnessUpdated,
    } = initialState;

    expect(wellnessEntries).toEqual({});
    expect(wellnessUpdated).toEqual(false);
    expect(requestSetWellnessEntries({} as any)).toBeNull();
    expect(updateEntryByKey({} as any)).toBeNull();
    expect(resetWellnessUpdated()).toBeNull();
  });

  it('should render provider with children', () => {
    const { asFragment } = render(
      <WellnessEntriesContextProvider>
        <TestComponent />
      </WellnessEntriesContextProvider>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render with initial-state', () => {
    const { asFragment } = render(
      <WellnessEntriesContextProvider
        initialState={{
          ['02-02-2023']: {
            water: { value: 5 },
            excercise: { time: '00:35' },
          },
        }}
      >
        <MockInitialEntries />
      </WellnessEntriesContextProvider>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should handle null initial state', () => {
    const { asFragment } = render(
      <WellnessEntriesContextProvider initialState={null as any}>
        <MockInitialEntries />
      </WellnessEntriesContextProvider>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render initial entries', () => {
    const { asFragment } = render(
      <WellnessEntriesContextProvider>
        <MockInitialEntries />
      </WellnessEntriesContextProvider>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should set entries', async () => {
    const { asFragment } = render(
      <WellnessEntriesContextProvider>
        <MockInitialEntries setEntries />
      </WellnessEntriesContextProvider>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('should update entry for non-logged in user', async () => {
    mockUseUserContext.mockReturnValue({ userLoggedIn: false } as any);

    const { asFragment } = render(
      <WellnessEntriesContextProvider>
        <MockUpdateEntry />
      </WellnessEntriesContextProvider>,
    );

    expect(mockDiaryService.updateWellnessEntries).not.toHaveBeenCalled();
    expect(asFragment()).toMatchSnapshot();
  });

  it('should update entry for logged in user', async () => {
    mockUseUserContext.mockReturnValue({ userLoggedIn: true } as any);

    const { asFragment } = render(
      <WellnessEntriesContextProvider>
        <MockUpdateEntry />
      </WellnessEntriesContextProvider>,
    );

    await act(() => Promise.resolve());
    jest.advanceTimersByTime(1000);

    expect(mockDiaryService.updateWellnessEntries).toHaveBeenCalledWith({
      body: { '01-02-2023': { date: '01-02-2023', water: { value: 4 } } },
    });
    expect(asFragment()).toMatchSnapshot();
  });

  it('should handle entry added to same date and not re-add date and allow for reset entries', async () => {
    mockUseUserContext.mockReturnValue({ userLoggedIn: true } as any);

    render(
      <WellnessEntriesContextProvider>
        <MockUpdateEntry runBothUpdates runResetWellnessUpdated />
      </WellnessEntriesContextProvider>,
    );

    await act(() => Promise.resolve());
    jest.advanceTimersByTime(1000);

    expect(mockDiaryService.updateWellnessEntries).toHaveBeenCalledWith({
      body: { '01-02-2023': { alcohol: { value: 7 }, date: '01-02-2023' } },
    });
  });

  it('should catch errors from updateWellnessEntries request', async () => {
    mockUseUserContext.mockReturnValue({ userLoggedIn: true } as any);
    mockDiaryService.updateWellnessEntries.mockResolvedValue({
      error: 'mock error occurred',
    });

    render(
      <WellnessEntriesContextProvider>
        <MockUpdateEntry />
      </WellnessEntriesContextProvider>,
    );

    await act(() => Promise.resolve());
    jest.advanceTimersByTime(1000);
    await act(() => Promise.resolve());

    expect(mockDiaryService.updateWellnessEntries).toHaveBeenCalled();
    expect(mockError).toHaveBeenCalledWith(
      'Error updating wellness entries:',
      'mock error occurred',
    );
  });
});
