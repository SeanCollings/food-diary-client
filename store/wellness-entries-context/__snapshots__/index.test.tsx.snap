// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`store - wellness-entries-context should handle null initial state 1`] = `
<DocumentFragment>
  <div>
    wellnessUpdated: false
  </div>
</DocumentFragment>
`;

exports[`store - wellness-entries-context should render initial entries 1`] = `
<DocumentFragment>
  <div>
    wellnessUpdated: false
  </div>
</DocumentFragment>
`;

exports[`store - wellness-entries-context should render provider with children 1`] = `
<DocumentFragment>
  <div>
    Test component
  </div>
</DocumentFragment>
`;

exports[`store - wellness-entries-context should render with initial-state 1`] = `
<DocumentFragment>
  <div>
    Key: 02-02-2023
    <div>
      {"water":{"value":5},"excercise":{"time":"00:35"}}
    </div>
  </div>
  <div>
    wellnessUpdated: false
  </div>
</DocumentFragment>
`;

exports[`store - wellness-entries-context should set entries 1`] = `
<DocumentFragment>
  <div>
    Key: 01-02-2023
    <div>
      {"water":{"value":5},"tea_coffee":{"value":2},"alcohol":{"value":1},"excercise":{"time":"00:15","details":"Run and walk"}}
    </div>
  </div>
  <div>
    wellnessUpdated: false
  </div>
</DocumentFragment>
`;

exports[`store - wellness-entries-context should update entry for logged in user 1`] = `
<DocumentFragment>
  <div>
    Key: 01-02-2023
    <div>
      {"water":{"value":4}}
    </div>
  </div>
  <div>
    wellnessUpdated: false
  </div>
</DocumentFragment>
`;

exports[`store - wellness-entries-context should update entry for non-logged in user 1`] = `
<DocumentFragment>
  <div>
    Key: 01-02-2023
    <div>
      {"water":{"value":4}}
    </div>
  </div>
  <div>
    wellnessUpdated: false
  </div>
</DocumentFragment>
`;
